contracts/LiquidityConversionRates.sol

```diff
@@ -1,57 +1,249 @@
+            uint currentBlockNumber,
+            bool buy,
+            uint qtyInSrcWei
+    ) public view returns(uint) {
 
+        uint rateInPRECISION;
         currentBlockNumber;
 
-        if(buy) {
-          // ETH goes in, token goes out
-          deltaE = qty * precision / 10**ETH_DECIMALS;
-          if(deltaE == 0) return 0;
+        require(qtyInSrcWei < MAX_QTY);
+        uint eInFp = fromWeiToFp(reserveContract.balance);
+        rateInPRECISION = getRateWithE(conversionToken, buy, qtyInSrcWei, eInFp);
+        require(rateInPRECISION < MAX_RATE);
+        return rateInPRECISION;
+    }
+
+    function recordImbalance(
+        ERC20 conversionToken,
+        int buyAmountInTwei,
+        uint rateUpdateBlock,
+        uint currentBlock
+    )
+        public
+    {
+        conversionToken;
+        rateUpdateBlock;
+        currentBlock;
+
+        require(msg.sender == reserveContract);
+        collectedFeesInTwei += calcCollectedFee(abs(buyAmountInTwei));
+    }
+
+    event CollectedFeesReset(uint resetFeesInTwei);
 
-          deltaT = deltaTFunc(r,Pmin,E,deltaE,precision);
+    function resetCollectedFees() public onlyAdmin {
+        uint resetFeesInTwei = collectedFeesInTwei;
+        collectedFeesInTwei = 0;
 
-          rate = deltaT * PRECISION / deltaE;
+        CollectedFeesReset(resetFeesInTwei);
+    }
+
+    function getRateWithE(ERC20 conversionToken, bool buy, uint qtyInSrcWei, uint eInFp) public view returns(uint) {
+        uint deltaEInFp;
+        uint deltaTInFp;
+        uint rateInPRECISION;
+
+        require(qtyInSrcWei < MAX_QTY);
+        require(eInFp < maxQtyInFp);
+
+        if (conversionToken != token) return 0;
+
+        if (buy) {
+            // ETH goes in, token goes out
+            deltaEInFp = fromWeiToFp(qtyInSrcWei);
+            if (deltaEInFp > maxCapBuyInFp) return 0;
+
+            if (deltaEInFp == 0) {
+                rateInPRECISION = buyRateZeroQuantity(eInFp);
+            } else {
+                rateInPRECISION = buyRate(eInFp, deltaEInFp);
+            }
+        } else {
+            deltaTInFp = fromTweiToFp(qtyInSrcWei);
+            deltaTInFp = reduceFee(deltaTInFp);
+            if (deltaTInFp == 0) {
+                rateInPRECISION = sellRateZeroQuantity(eInFp);
+                deltaEInFp = 0;
+            } else {
+                (rateInPRECISION, deltaEInFp) = sellRate(eInFp, deltaTInFp);
+            }
+
+            if (deltaEInFp > maxCapSellInFp) return 0;
         }
-        else {
-          deltaT = qty * precision / (10** getDecimals(token));
-          if(deltaT == 0) return 0;
 
-          deltaE = deltaEFunc(r,Pmin,E,deltaT,precision,numPrecisionBits);
-          rate = deltaE * PRECISION / deltaT;
+        rateInPRECISION = validateRate(rateInPRECISION, buy);
+        require(rateInPRECISION < MAX_RATE);
+        return rateInPRECISION;
+    }
+
+    function validateRate(uint rateInPRECISION, bool buy) public view returns(uint) {
```

Comment:
name of function is bad.
It is not boolean validation.
Please find better name.